src/sim/kernel_workload.cc

```diff
@@ -104,12 +104,28 @@ KernelWorkload::initState()
     };
     if (params().object_file != "")  {
         if (params().addr_check) {
+            auto memory_size = system->memSize();
+            auto kernel_size = _end - _start;
             // Validate kernel mapping before loading binary
+            char error_msg_buffer [500];
+
+            if (kernel_size > memory_size){
+                snprintf(error_msg_buffer, 500, "Kernel is mapped to invalid "
+                "location (not memory). start (%#lx) - end (%#lx) %#lx:%#lx\n"
+                "The kernel size (%ld bytes, %ld MB) appears to be larger "
+                "than total system memory (%ld bytes, %ld MB). Try increasing"
+                " system memory.", _start, _end, mapper(_start), mapper(_end),
+                kernel_size, kernel_size>>20, memory_size, memory_size>>20);
+            }
+            else{
+                snprintf(error_msg_buffer, 500, "Kernel is mapped to invalid"
+                "location (not memory). start (%#lx) - end (%#lx) %#lx:%#lx\n",
+                    _start, _end, mapper(_start), mapper(_end));
+            }
```

Comment:
This seems to me a bit too convoluted... First we populate the error message and then we print it but only if the fatal condition is met? This means the second snprintf will happen even if there is no issue whatsoever. I prefer Bobby solution of handling everything with ternary operator... Alternatively I guess am fine on having 2 fatal_ifs